### PR TITLE
[Improve] Enhance complement function transformation 

### LINKED ISSUE #10370: [Improvement] [Api & Master] Complement function transformation
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

The function of replenishment is changed to that the user enters the data date of the amount to be replenished. The system generates a batch of task instances in batch based on the date entered by the user

### What you expected to happen

Remove the dependency on scheduled tasks

### How to reproduce

1.The replenishment function is changed to enter the data date of the amount to be replenished by the user. The system generates a batch of task instances in batch based on the date entered by the user.
2.Remove the dependency on scheduled tasks
3.Change the filling method of data date to manual input of text box
image

### Anything else

_No response_

### Version

dev

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by hstdream**
close #10370

### CODE REVIEW

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/Constants.java [hunk 1]
```diff
@@ -361,6 +361,11 @@ private Constants() {
      */
     public static final String CMDPARAM_COMPLEMENT_DATA_END_DATE = "complementEndDate";
 
+    /**
+     * complement data Schedule date
+     */
+    public static final String CMDPARAM_COMPLEMENT_DATA_SCHEDULE_DATE = "complementScheduleDate";
```
Comment (lenboo) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/Constants.java]: It would be a list string.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/corn/CronUtils.java [hunk 2]
```diff
@@ -35,13 +34,9 @@
 import org.apache.commons.collections.CollectionUtils;
 
 import java.text.ParseException;
-import java.util.ArrayList;
-import java.util.Calendar;
-import java.util.Collections;
-import java.util.Date;
-import java.util.GregorianCalendar;
-import java.util.List;
+import java.util.*;
```
Comment (SbloodyS) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/corn/CronUtils.java]: Same here.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java [hunk 3]
```diff
@@ -18,13 +18,9 @@
 package org.apache.dolphinscheduler.api.service.impl;
 
 import static org.apache.dolphinscheduler.api.constants.ApiFuncIdentificationConstant.WORKFLOW_START;
-import static org.apache.dolphinscheduler.common.Constants.CMDPARAM_COMPLEMENT_DATA_END_DATE;
-import static org.apache.dolphinscheduler.common.Constants.CMDPARAM_COMPLEMENT_DATA_START_DATE;
-import static org.apache.dolphinscheduler.common.Constants.CMD_PARAM_RECOVER_PROCESS_ID_STRING;
-import static org.apache.dolphinscheduler.common.Constants.CMD_PARAM_START_NODES;
-import static org.apache.dolphinscheduler.common.Constants.CMD_PARAM_START_PARAMS;
-import static org.apache.dolphinscheduler.common.Constants.MAX_TASK_TIMEOUT;
+import static org.apache.dolphinscheduler.common.Constants.*;
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: Please avoid import *

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java [hunk 4]
```diff
@@ -75,13 +71,7 @@
...
 import org.apache.commons.lang3.StringUtils;
 
-import java.util.ArrayList;
-import java.util.Date;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: Same here.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java [hunk 5]
```diff
@@ -17,12 +17,7 @@
 
 package org.apache.dolphinscheduler.server.master.runner;
 
-import static org.apache.dolphinscheduler.common.Constants.CMDPARAM_COMPLEMENT_DATA_END_DATE;
-import static org.apache.dolphinscheduler.common.Constants.CMDPARAM_COMPLEMENT_DATA_START_DATE;
-import static org.apache.dolphinscheduler.common.Constants.CMD_PARAM_RECOVERY_START_NODE_STRING;
-import static org.apache.dolphinscheduler.common.Constants.CMD_PARAM_RECOVER_PROCESS_ID_STRING;
-import static org.apache.dolphinscheduler.common.Constants.CMD_PARAM_START_NODES;
-import static org.apache.dolphinscheduler.common.Constants.DEFAULT_WORKER_GROUP;
+import static org.apache.dolphinscheduler.common.Constants.*;
```
Comment (SbloodyS) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java]: Same here.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java [hunk 6]
```diff
@@ -740,45 +729,24 @@ protected int createComplementCommandList(Date start, Date end, RunMode runMode,
...
+                        cmdParam.put(CMDPARAM_COMPLEMENT_DATA_SCHEDULE_DATE, tempDate.substring(0,tempDate.length() - 1));
                         command.setCommandParam(JSONUtils.toJsonString(cmdParam));
                         processService.createCommand(command);
-
-                        if (schedules.isEmpty() || complementDependentMode == ComplementDependentMode.OFF_MODE) {
-                            logger.info("process code: {} complement dependent in off mode or schedule's size is 0, skip "
-                                    + "dependent complement data", command.getProcessDefinitionCode());
-                        } else {
-                            dependentProcessDefinitionCreateCount += createComplementDependentCommand(schedules, command);
-                        }
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: Why did you remove this? It may cause complement dependent process error.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java [hunk 7]
```diff
@@ -740,45 +729,24 @@ protected int createComplementCommandList(Date start, Date end, RunMode runMode,
...
-
-                        if (i == 1) {
-                            endDateIndex += singleCommandItems - 1;
-                        } else {
-                            startDateIndex = endDateIndex + 1;
-                            endDateIndex += singleCommandItems;
+                    for(List<String> stringDate : Lists.partition(listDate,createCount)){
+                        String tempDate = "";
+                        for(String date : stringDate){
+                            tempDate  += date +",";
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: It's recommanded to format this line.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java [hunk 8]
```diff
@@ -740,45 +729,24 @@ protected int createComplementCommandList(Date start, Date end, RunMode runMode,
...
                         }
-
-                        cmdParam.put(CMDPARAM_COMPLEMENT_DATA_START_DATE, DateUtils.dateToString(listDate.get(startDateIndex)));
-                        cmdParam.put(CMDPARAM_COMPLEMENT_DATA_END_DATE, DateUtils.dateToString(listDate.get(endDateIndex)));
+                        cmdParam.put(CMDPARAM_COMPLEMENT_DATA_SCHEDULE_DATE, tempDate.substring(0,tempDate.length() - 1));
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: I think using it's a little confusing using ```substring``` to remove the last comma. Is it more elegant to use ```String.join(",", tmpDateList)``` ?

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java [hunk 9]
```diff
@@ -943,15 +938,12 @@ private void initTaskQueue() {
...
-            if (cmdParam != null && cmdParam.containsKey(CMDPARAM_COMPLEMENT_DATA_START_DATE)) {
+            if (cmdParam != null && cmdParam.containsKey(CMDPARAM_COMPLEMENT_DATA_SCHEDULE_DATE)) {
                 // reset global params while there are start parameters
                 setGlobalParamIfCommanded(processDefinition, cmdParam);
 
-                Date start = DateUtils.stringToDate(cmdParam.get(CMDPARAM_COMPLEMENT_DATA_START_DATE));
-                Date end = DateUtils.stringToDate(cmdParam.get(CMDPARAM_COMPLEMENT_DATA_END_DATE));
-                List<Schedule> schedules = processService.queryReleaseSchedulerListByProcessDefinitionCode(processInstance.getProcessDefinitionCode());
                 if (complementListDate.size() == 0 && needComplementProcess()) {
-                    complementListDate = CronUtils.getSelfFireDateList(start, end, schedules);
```
Comment (SbloodyS) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java]: As we discuss in wechat. I do not think this is a good idea to remove such a big feature of complement data.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/enums/Status.java [hunk 10]
```diff
@@ -407,6 +407,8 @@ public enum Status {
 
     NO_CURRENT_OPERATING_PERMISSION(1400001, "The current user does not have this permission.", "当前用户无此权限"),
     FUNCTION_DISABLED(1400002, "The current feature is disabled.", "当前功能已被禁用"),
+    SCHEDULE_TIME_NUMBER(1400003, "The number of complement dates exceed 100.", "补数日期个数超过100"),
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/enums/Status.java:410]: I think we should not limit the date range of user complement. In the actual production environment, it is very common to supplement the time range within one or two years.

**Review by SbloodyS — CHANGES_REQUESTED** [review]

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java (hunk 6 above)
Comment (caishunfeng) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: +1, I had the same question here.

**Review by caishunfeng — COMMENTED** [review]
I think it's not a complete pr and we should add the auto generate date tool on ui.

**Comment by lenboo** [comment]
I think we can keep two complement data way in the same time.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/enums/Status.java (hunk 10 above)
Comment (hstdream) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/enums/Status.java:410]: The current solution is to manually enter and use the selected date. Manual input requires limiting the number and length of the entire set

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java (hunk 7 above)
Comment (hstdream) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: 1

**Comment by hstdream** [comment]
@SbloodyS @caishunfeng

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/Constants.java (hunk 1 above)
Comment (hstdream) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/Constants.java]: Modified

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/corn/CronUtils.java (hunk 2 above)
Comment (hstdream) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/corn/CronUtils.java]: Modified

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java (hunk 3 above)
Comment (hstdream) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: OK，Modified

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java (hunk 6 above)
Comment (hstdream) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: OK，Modified

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java (hunk 8 above)
Comment (hstdream) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: OK，Modified

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java (hunk 9 above)
Comment (hstdream) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java]: 目前已经修改成手动输入和自动选择两种方式
At present, it has been modified to manual input and automatic selection

Comment (hstdream) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java]: At present, it has been modified to manual input and automatic selection

File: dolphinscheduler-e2e/dolphinscheduler-e2e-case/src/test/java/org/apache/dolphinscheduler/e2e/cases/WorkflowE2ETest.java [hunk 11]
```diff
@@ -162,56 +162,56 @@ void testCreateSubWorkflow() {
...
+//            final Row row = projectPage
+//                    .goToTab(WorkflowInstanceTab.class)
+//                    .instances()
+//                    .iterator()
+//                    .next();
+//
+//            assertThat(row.isSuccess()).isTrue();
+//            assertThat(row.executionTime()).isEqualTo(2);
+//        });
+//    }
```
Comment (SbloodyS) [inline dolphinscheduler-e2e/dolphinscheduler-e2e-case/src/test/java/org/apache/dolphinscheduler/e2e/cases/WorkflowE2ETest.java]: Why did you comment out this?

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/corn/CronUtils.java (hunk 2 above)
Comment (SbloodyS) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/corn/CronUtils.java]: This has not been addressed.

File: dolphinscheduler-e2e/dolphinscheduler-e2e-case/src/test/java/org/apache/dolphinscheduler/e2e/cases/WorkflowE2ETest.java (hunk 11 above)
Comment (hstdream) [reply dolphinscheduler-e2e/dolphinscheduler-e2e-case/src/test/java/org/apache/dolphinscheduler/e2e/cases/WorkflowE2ETest.java]: Because the input parameters of runworkflow are modified, the E2E is shielded for the time being. After the front-end code is merged, it is opened again.

Comment (SbloodyS) [reply dolphinscheduler-e2e/dolphinscheduler-e2e-case/src/test/java/org/apache/dolphinscheduler/e2e/cases/WorkflowE2ETest.java]: Currently our E2E test does not cover the timing module and the complement module. I think we should not change it.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/corn/CronUtils.java [hunk 12]
```diff
@@ -16,14 +16,13 @@
  */
 
 package org.apache.dolphinscheduler.service.corn;
-
+import static org.apache.dolphinscheduler.common.Constants.CMDPARAM_COMPLEMENT_DATA_SCHEDULE_DATE_LIST;
```
Comment (SbloodyS) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/corn/CronUtils.java]: There needs to be a blank line in front of this according to checkstyle.

Comment (hstdream) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/corn/CronUtils.java]: ok

**Review by WangJPLeo — DISMISSED** [review]
+1

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java (hunk 4 above)
Comment (SbloodyS) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: This has not been addressed.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java [hunk 13]
```diff
@@ -187,11 +183,20 @@ public Map<String, Object> execProcessInstance(User loginUser, long projectCode,
             return result;
         }
 
+        if(!checkScheduleTimeNum(commandType,cronTime)){
+            putMsg(result, Status.SCHEDULE_TIME_NUMBER);
+            return result;
+        }
+
+        if(!checkScheduleTimeRepeat(commandType,cronTime)){
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: I think we should delete duplicate data here instead of return error. This will improve the user experience.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java [hunk 14]
```diff
@@ -686,89 +728,116 @@ private int createCommand(CommandType commandType, long processDefineCode,
...
+        if(scheduleParam.containsKey(CMDPARAM_COMPLEMENT_DATA_START_DATE) && scheduleParam.containsKey(CMDPARAM_COMPLEMENT_DATA_END_DATE)){
+            startDate = scheduleParam.get(CMDPARAM_COMPLEMENT_DATA_START_DATE);
+            endDate = scheduleParam.get(CMDPARAM_COMPLEMENT_DATA_END_DATE);
+        }
         switch (runMode) {
             case RUN_MODE_SERIAL: {
-                if (start.after(end)) {
-                    logger.warn("The startDate {} is later than the endDate {}", start, end);
-                    break;
+                if(StringUtils.isNotEmpty(dateList) || dateList != null){
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: I think ```StringUtils.isNotEmpty(null)``` would return false. So it's duplicated.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java [hunk 15]
```diff
@@ -686,89 +728,116 @@ private int createCommand(CommandType commandType, long processDefineCode,
...
+                    cmdParam.put(CMDPARAM_COMPLEMENT_DATA_SCHEDULE_DATE_LIST,dateList);
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: This line should be formated.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java [hunk 16]
```diff
@@ -686,89 +728,116 @@ private int createCommand(CommandType commandType, long processDefineCode,
...
-                        int extra = (i <= remainingItems) ? 1 : 0;
-                        int singleCommandItems = (itemsPerCommand + extra);
-
-                        if (i == 1) {
-                            endDateIndex += singleCommandItems - 1;
-                        } else {
-                            startDateIndex = endDateIndex + 1;
-                            endDateIndex += singleCommandItems;
+                }
+                if(StringUtils.isNotEmpty(dateList) || dateList != null){
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: Same as 754 line.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java [hunk 17]
```diff
@@ -854,4 +923,42 @@ private List<DependentProcessDefinition> checkDependentProcessDefinitionValid(Li
...
+            if (startDate == null || endDate == null) {
+              return 0;
+            }
+            start = DateUtils.getScheduleDate(startDate);
+            end = DateUtils.getScheduleDate(endDate);
+            if(start == null || end == null){
+                return 0;
+            }
+            if (start.after(end)) {
+                logger.info("complement data error, wrong date start:{} and end date:{} ",
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: This should be ```logger.error```

**Review by SbloodyS — COMMENTED** [review]
I let some comment below. BTW, we should put this feature into docs.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java [hunk 18]
```diff
@@ -228,6 +233,54 @@ private boolean checkMasterExists(Map<String, Object> result) {
         return true;
     }
 
+    /**
+     *
+     * @param complementData
+     * @param cronTime
+     * @return CommandType is COMPLEMENT_DATA and cronTime's number is not greater than  return true , otherwise return false
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: Missing greater than 100.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ExecutorController.java [hunk 19]
```diff
@@ -88,6 +88,9 @@ public class ExecutorController extends BaseController {
      * @param projectCode project code
      * @param processDefinitionCode process definition code
      * @param scheduleTime schedule time
+     *                     when CommandType is COMPLEMENT_DATA  there are two ways to transfer parameters
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ExecutorController.java]: It's recommand to comment before of after the `@param` instead of between them.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java (hunk 14 above)
Comment (hstdream) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: ok

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java (hunk 17 above)
Comment (hstdream) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: ok

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ExecutorController.java (hunk 19 above)
Comment (hstdream) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ExecutorController.java]: If the description is long, do you need to wrap

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java (hunk 13 above)
Comment (hstdream) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: ok

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java (hunk 4 above)
Comment (zhongjiajie) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: why our codesytle check can not catch this?

**Comment by zhongjiajie** [comment]
PTAL at this PR if you have time @SbloodyS

File: docs/docs/en/about/glossary.md [hunk 20]
```diff
@@ -45,7 +45,7 @@ provided. **Continue** refers to regardless of the status of the task running in
 failure. **End** means that once a failed task is found, Kill will also run the parallel task at the same time, and the
 process fails and ends
 
-**Complement**: Supplement historical data，Supports **interval parallel and serial** two complement methods
+**Complement**: Supplement historical data，Supports **interval parallel and serial** two complement methods, and supports **select date range and manually enter date** two date methods
```
Comment (zhongjiajie) [inline docs/docs/en/about/glossary.md]:
```suggestion
**Complement**: Supplement historical data，supports **interval parallel and serial** two complement methods, and two types of date selection which include **date range** and **date enumeration **.
```

File: docs/docs/zh/about/glossary.md [hunk 21]
```diff
@@ -30,7 +30,7 @@
 
 **失败策略**：对于并行运行的任务，如果有任务失败，提供两种失败策略处理方式，**继续**是指不管并行运行任务的状态，直到流程失败结束。**结束**是指一旦发现失败任务，则同时Kill掉正在运行的并行任务，流程失败结束
 
-**补数**：补历史数据，支持**区间并行和串行**两种补数方式
+**补数**：补历史数据，支持**区间并行和串行**两种补数方式，并且支持**选择日期范围和手动输入日期**两种日期方式
```
Comment (zhongjiajie) [inline docs/docs/zh/about/glossary.md]:
```suggestion
**补数**：补历史数据，支持**区间并行和串行**两种补数方式，其日期选择方式包括**日期范围**和**日期枚举**两种
```

File: docs/docs/en/about/glossary.md [hunk 22]
```diff
@@ -45,7 +45,7 @@ provided. **Continue** refers to regardless of the status of the task running in
 failure. **End** means that once a failed task is found, Kill will also run the parallel task at the same time, and the
 process fails and ends
 
-**Complement**: Supplement historical data，Supports **interval parallel and serial** two complement methods
+**Complement**: Supplement historical data，supports **interval parallel and serial** two complement methods, and two types of date selection which include **date range** and **date enumeration **.
```
Comment (zhongjiajie) [inline docs/docs/en/about/glossary.md]:
```suggestion
**Complement**: Supplement historical data，supports **interval parallel and serial** two complement methods, and two types of date selection which include **date range** and **date enumeration**.
```

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/Constants.java [hunk 23]
```diff
@@ -809,4 +814,10 @@ private Constants() {
      * tenant
      */
     public static final int TENANT_FULL_NAME_MAX_LENGTH = 30;
+
+    /**
+     * schedule time
+     */
+    public static final int SCHEDULE_TIME_MAX_LENGTH = 100;
```
Comment (zhongjiajie) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/Constants.java:821]: Maybe we should add some docs about why we should set this value to `100` instead of `200` or other values

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java [hunk 24]
```diff
@@ -774,7 +775,12 @@ private int createComplementDataCommand(Date scheduleDate) {
         if (cmdParam.containsKey(Constants.CMD_PARAM_RECOVERY_START_NODE_STRING)) {
             cmdParam.remove(Constants.CMD_PARAM_RECOVERY_START_NODE_STRING);
         }
-        cmdParam.replace(CMDPARAM_COMPLEMENT_DATA_START_DATE, DateUtils.format(scheduleDate, "yyyy-MM-dd HH:mm:ss", null));
+        if(cmdParam.containsKey(CMDPARAM_COMPLEMENT_DATA_SCHEDULE_DATE_LIST)){
+            cmdParam.replace(CMDPARAM_COMPLEMENT_DATA_SCHEDULE_DATE_LIST, cmdParam.get(CMDPARAM_COMPLEMENT_DATA_SCHEDULE_DATE_LIST).substring(cmdParam.get(CMDPARAM_COMPLEMENT_DATA_SCHEDULE_DATE_LIST).indexOf(",")+1));
+        }
+        if(cmdParam.containsKey(CMDPARAM_COMPLEMENT_DATA_START_DATE)){
+            cmdParam.replace(CMDPARAM_COMPLEMENT_DATA_START_DATE, DateUtils.format(scheduleDate, "yyyy-MM-dd HH:mm:ss", null));
```
Comment (zhongjiajie) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java]: I think we have `Constants` value with both `","` and `"yyyy-MM-dd HH:mm:ss"`, should avoid using bare common stirng in our code

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/Constants.java (hunk 23 above)
Comment (hstdream) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/Constants.java:821]: This figure is neither too much nor too little. It is tentatively set at 100 at present

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java (hunk 24 above)
Comment (hstdream) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java]: Thanks,I also think

**Comment by hstdream** [comment]
docs/docs/zh/guide/project/workflow-definition.md  this docs supplement after the front-end code is updated

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java (hunk 4 above)
Comment (SbloodyS) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java]: > why our codesytle check can not catch this?

I'm also confusing about it.

**Review by SbloodyS — COMMENTED** [review]
LGTM overall. But we still have discussion in https://github.com/apache/dolphinscheduler/pull/10376#discussion_r894520852

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java [hunk 25]
```diff
@@ -686,89 +705,117 @@ private int createCommand(CommandType commandType, long processDefineCode,
...
-                List<Schedule> schedules = processService.queryReleaseSchedulerListByProcessDefinitionCode(command.getProcessDefinitionCode());
-                listDate.addAll(CronUtils.getSelfFireDateList(start, end, schedules));
-                int listDateSize = listDate.size();
-                createCount = listDate.size();
-                if (!CollectionUtils.isEmpty(listDate)) {
-                    if (expectedParallelismNumber != null && expectedParallelismNumber != 0) {
-                        createCount = Math.min(listDate.size(), expectedParallelismNumber);
-                        if (listDateSize < createCount) {
-                            createCount = listDateSize;
+                if(startDate != null && endDate != null){
```
Comment (caishunfeng) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java:755]: It's better to separate these out into a method like: createComplementCommandByCron and createComplementCommandByDateList, just a code improvement suggestion.

**Comment by caishunfeng** [comment]
> LGTM overall. But we still have discussion in [#10376 (comment)](https://github.com/apache/dolphinscheduler/pull/10376#discussion_r894520852)

Hi @SbloodyS you should give a approve because the change requested.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java (hunk 25 above)
Comment (hstdream) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java:755]: Thanks,here I will consider how to improvement

**Comment by SbloodyS** [comment]
> > LGTM overall. But we still have discussion in [#10376 (comment)](https://github.com/apache/dolphinscheduler/pull/10376#discussion_r894520852)
> 
> Hi @SbloodyS you should give a approve because the change requested.

Sorry. I don't have strong opinion here. I just not quite sure about whether should give a approve since we still have unresolved discussions.

I gave approved. Do you have better suggestions? @zhongjiajie @caishunfeng

**Review by SbloodyS — DISMISSED** [review]

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java (hunk 24 above)
Comment (caishunfeng) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java]: This discussion is not resolved

**Comment by caishunfeng** [comment]
> > > LGTM overall. But we still have discussion in [#10376 (comment)](https://github.com/apache/dolphinscheduler/pull/10376#discussion_r894520852)
> > 
> > 
> > Hi @SbloodyS you should give a approve because the change requested.
> 
> Sorry. I don't have strong opinion here. I just not quite sure about whether should give a approve since we still have unresolved discussions.
> 
> I gave approved. Do you have better suggestions? @zhongjiajie @caishunfeng

Sorry, I got it wrong, thanks for pointing it out.

**Review by caishunfeng — DISMISSED** [review]
+1

**Comment by zhongjiajie** [comment]
> > > LGTM overall. But we still have discussion in [#10376 (comment)](https://github.com/apache/dolphinscheduler/pull/10376#discussion_r894520852)
> > 
> > Hi @SbloodyS you should give a approve because the change requested.
> 
> Sorry. I don't have strong opinion here. I just not quite sure about whether should give a approve since we still have unresolved discussions.
> 
> I gave approved. Do you have better suggestions? @zhongjiajie @caishunfeng 

i think we better add some java doc on that constants value about why we set it to 100 instead of other value

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java (hunk 24 above)
Comment (zhongjiajie) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java]: it seem that you still not change this

**Comment by hstdream** [comment]
> > > > LGTM overall. But we still have discussion in [#10376 (comment)](https://github.com/apache/dolphinscheduler/pull/10376#discussion_r894520852)
> > > 
> > > 
> > > Hi @SbloodyS you should give a approve because the change requested.
> > 
> > 
> > Sorry. I don't have strong opinion here. I just not quite sure about whether should give a approve since we still have unresolved discussions.
> > I gave approved. Do you have better suggestions? @zhongjiajie @caishunfeng
> 
> i think we better add some java doc on that constants value about why we set it to 100 instead of other value
At that time, it was considered that the amount of data was too large, and then set the value of 100.

**Review by zhongjiajie — DISMISSED** [review]
LGTM, waittig CI

**Comment by zhongjiajie** [comment]
I restart the failing `E2E/fileManage`CI, and why this test failed often

**Review by zhongjiajie — APPROVED** [review]
